File: modules/dose_recommendation.R
# modules/dose_recommendation.R

library(shiny)
library(DT)
library(ggplot2)
library(nlmixr2)

# 剂量推荐的 UI
doseRecommendationUI <- function(id) {
  ns <- NS(id)
  tagList(
    box(title = "剂量推荐", status = "primary", solidHeader = TRUE,
        selectInput(ns("selected_model"), "选择模型结果",
                    choices = NULL),
        numericInput(ns("desired_dv"), "期望药物浓度 (DV)", value = 10, min = 0),
        numericInput(ns("loading_dose"), "加载剂量 (AMT)", value = 100, min = 0),
        numericInput(ns("maintenance_dose"), "维持剂量 (AMT)", value = 50, min = 0),
        numericInput(ns("interval"), "给药间隔 (小时)", value = 12, min = 1),
        actionButton(ns("recommend_dose_button"), "推荐剂量")
    ),
    box(title = "推荐剂量结果", status = "info", solidHeader = TRUE,
        DT::dataTableOutput(ns("dose_table"))
    ),
    box(title = "剂量推荐图表", status = "warning", solidHeader = TRUE,
        plotOutput(ns("dose_plot"), height = "400px")
    )
  )
}

# 剂量推荐的服务器逻辑
doseRecommendationServer <- function(id) {
  moduleServer(
    id,
    function(input, output, session) {
      
      # 获取模型结果文件列表
      model_files <- reactive({
        if (!dir.exists("PKObjResultsFolder")) {
          dir.create("PKObjResultsFolder")
        }
        files <- list.files("PKObjResultsFolder", pattern = "\\.ressum\\.rds$", full.names = FALSE)
        return(files)
      })
      
      # 更新选择模型结果的下拉菜单
      observe({
        updateSelectInput(session, "selected_model", choices = model_files())
      })
      
      # 剂量推荐逻辑
      dose_data <- eventReactive(input$recommend_dose_button, {
        req(input$selected_model, input$desired_dv, input$loading_dose, input$maintenance_dose, input$interval)
        
        file_path <- file.path("PKObjResultsFolder", input$selected_model)
        if (file.exists(file_path)) {
          res_sum <- readRDS(file_path)
          
          # 假设 res_sum 包含拟合后的模型对象
          if (!is.null(res_sum$model_fit)) {
            fit <- res_sum$model_fit
            
            # 定义模拟函数
            simulate_dose <- function(amt, interval, times) {
              dosing <- data.frame(AMT = amt, TIME = seq(0, times, by = interval))
              simulate(fit, newdata = dosing)
            }
            
            # 运行模拟
            simulation_time <- 72  # 模拟总时间（小时）
            sim_data <- simulate_dose(input$loading_dose, input$interval, simulation_time)
            
            # 计算平均浓度
            avg_dv <- mean(sim_data$DV, na.rm = TRUE)
            
            # 简单剂量调整逻辑
            recommended_dose <- input$maintenance_dose * (input$desired_dv / avg_dv)
            recommended_dose <- round(recommended_dose, 2)
            
            # 返回结果
            dose_result <- data.frame(
              加载剂量 = input$loading_dose,
              维持剂量 = recommended_dose,
              给药间隔 = input$interval,
              期望DV = input$desired_dv,
              模拟平均DV = round(avg_dv, 2)
            )
            
            # 绘制剂量推荐图表
            dose_plot <- ggplot(sim_data, aes(x = TIME, y = DV)) +
              geom_line(color = "blue") +
              geom_hline(yintercept = input$desired_dv, linetype = "dashed", color = "red") +
              theme_minimal() +
              labs(title = "剂量推荐模拟结果",
                   x = "时间 (小时)",
                   y = "药物浓度 (DV)")
            
            return(list(
              dose_table = dose_result,
              dose_plot = dose_plot
            ))
            
          } else {
            showModal(modalDialog(
              title = "错误",
              "选定的模型结果中未找到拟合后的模型对象。",
              easyClose = TRUE,
              footer = NULL
            ))
            return(NULL)
          }
        } else {
          showModal(modalDialog(
            title = "错误",
            "选定的模型结果文件不存在。",
            easyClose = TRUE,
            footer = NULL
          ))
          return(NULL)
        }
      })
      
      # 渲染推荐剂量表格
      output$dose_table <- DT::renderDataTable({
        data <- dose_data()
        req(data$dose_table)
        datatable(data$dose_table, options = list(pageLength = 5, scrollX = TRUE))
      })
      
      # 渲染剂量推荐图表
      output$dose_plot <- renderPlot({
        data <- dose_data()
        req(data$dose_plot)
        print(data$dose_plot)
      })
      
    }
  )
}

File: modules/dose_recommendation.py
# modules/dose_recommendation.py

from flask import Blueprint, render_template, request, jsonify, current_app
from flask_login import login_required
import os
import pandas as pd
import json
import numpy as np

dose_recommendation_bp = Blueprint('dose_recommendation', __name__, url_prefix='/dose_recommendation')

@dose_recommendation_bp.route('/', methods=['GET'])
@login_required
def recommendation_page():
    return render_template('recommendation.html')

@dose_recommendation_bp.route('/model_list', methods=['GET'])
@login_required
def model_list():
    try:
        # 获取PKObjResultsFolder文件夹中的模型结果文件
        if not os.path.exists('PKObjResultsFolder'):
            os.makedirs('PKObjResultsFolder')
        
        files = [f for f in os.listdir('PKObjResultsFolder') if f.endswith('.ressum.rds')]
        return jsonify({'status': 'success', 'model_files': files})
    except Exception as e:
        return jsonify({'status': 'error', 'message': str(e)})

@dose_recommendation_bp.route('/calculate', methods=['POST'])
@login_required
def calculate_dose():
    try:
        # 获取表单数据
        data = request.form
        selected_model = data.get('selected_model')
        desired_dv = float(data.get('desired_dv', 10))
        loading_dose = float(data.get('loading_dose', 100))
        maintenance_dose = float(data.get('maintenance_dose', 50))
        interval = float(data.get('interval', 12))
        
        # 简单剂量调整计算
        # 在实际应用中，你应该使用真实的PK模型
        recommended_dose = maintenance_dose * (desired_dv / (loading_dose * 0.12))
        recommended_dose = round(recommended_dose, 2)
        
        # 创建模拟数据用于绘图
        time = np.arange(0, 72, 0.5)
        simulated_concentration = []
        
        for t in time:
            dose_points = int(t / interval)
            conc = loading_dose * np.exp(-0.1 * (t % interval))
            
            for i in range(dose_points):
                conc += maintenance_dose * np.exp(-0.1 * ((t - interval * (i + 1)) % interval))
            
            simulated_concentration.append(conc)
        
        # 创建响应数据
        result = {
            'loading_dose': loading_dose,
            'maintenance_dose': recommended_dose,
            'interval': interval,
            'desired_dv': desired_dv,
            'simulated_dv': round(np.mean(simulated_concentration), 2)
        }
        
        # 创建绘图数据
        plot_data = {
            'traces': [
                {
                    'x': time.tolist(),
                    'y': simulated_concentration,
                    'mode': 'lines',
                    'name': '模拟浓度',
                    'line': {'color': 'rgb(55, 128, 191)', 'width': 3}
                },
                {
                    'x': [time[0], time[-1]],
                    'y': [desired_dv, desired_dv],
                    'mode': 'lines',
                    'name': '期望浓度',
                    'line': {'color': 'rgb(219, 64, 82)', 'width': 2, 'dash': 'dash'}
                }
            ],
            'layout': {
                'title': '剂量推荐模拟结果',
                'xaxis': {'title': '时间 (小时)'},
                'yaxis': {'title': '药物浓度 (DV)'}
            }
        }
        
        return jsonify({
            'status': 'success',
            'data': result,
            'plot_data': plot_data
        })
    
    except Exception as e:
        return jsonify({'status': 'error', 'message': str(e)})